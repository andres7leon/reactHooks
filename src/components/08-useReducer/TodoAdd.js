import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ {description}, handleInputChange, reset ] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {

        e.preventDefault();
 
        if( description.trim().length <= 1 ){
            return;
        }
        
        handleAddTodo({
            id: new Date().getTime(),
            desc: description,
            done: true
        });
 
        reset();
    }

    return (
        <>
            <h3>Form</h3>
            <hr/>
            <form onSubmit={ handleSubmit }>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender ..."
                    autoComplete="off"
                    onChange={ handleInputChange }
                    value={description}
                />
            
                <div className="d-grid gap-2"> 
                    <button 
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                    >
                        Agregar
                    </button>         
                </div>  
            </form>   
        </>
    )
}
