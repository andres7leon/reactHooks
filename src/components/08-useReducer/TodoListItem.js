import React from 'react'
import PropTypes from 'prop-types';

export const TodoListItem = ({ todos, handleToggle, handleRemove}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((todo,i)=>(
                <li key={ todo.id }
                className="list-group-item"                       
                >
                <p style={{cursor: 'pointer', textDecoration: todo.done ? 'line-through' : ''}} onClick={ ()=> { handleToggle(todo.id) } }>{ i + 1 }.  { todo.desc }</p>
                <button className="btn btn-danger" onClick={() => { handleRemove(todo.id) }}>
                    Borrar
                </button>
                </li> 
                ))
            } 
        </ul>
            
    )
}

TodoListItem.prototype = {
    todos: PropTypes.array
};
