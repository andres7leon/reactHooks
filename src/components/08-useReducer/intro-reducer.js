const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = ( state = initialState, action) => {

    if ( action?.type === 'agregar' ) {
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();


const newTodo = {
    id: 2,
    todo: 'Comprar Lechee',
    done: false
}

const actionAddProduct = {
    type: 'agregar',
    payload: newTodo
}

todos = todoReducer( todos, actionAddProduct);


console.log(todos)