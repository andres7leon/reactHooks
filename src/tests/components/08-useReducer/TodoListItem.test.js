import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demosTodo } from '../../fixtures/demoTodos';

describe('Pruebas Hook TodoListItem', () => {
    
    const handleRemove = jest.fn();
    const handleToggle = jest.fn();
    let wrapper = shallow( <TodoListItem todos={demosTodo} handleToggle={handleToggle} handleRemove={handleRemove} /> );

    test('debe de mostrarse bien', () => {
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('debe de correr la funcion de handleRemove', () => {
        const button = wrapper.find('button').at(0);
        button.simulate('click');
        expect( handleRemove ).toHaveBeenCalledWith( 1 );
    });
    
    test('debe de mostrarse el texto correctamente', () => {
        
        const p = wrapper.find('p').at(0);

        expect( p.text() ).toBe( `${demosTodo[0].id}.  ${demosTodo[0].desc}`)

    });

    test('se debe de completar', () => {
        
        demosTodo[0].done = true  

        wrapper = shallow( <TodoListItem todos={demosTodo} /> );
        
        const todoOk = wrapper.find('p').at(0).props().style?.textDecoration;

        expect( todoOk ).toBe('line-through');

    });
    

});
