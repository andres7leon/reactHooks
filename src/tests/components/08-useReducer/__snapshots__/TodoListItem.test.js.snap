// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas Hook TodoListItem debe de mostrarse bien 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item"
    key="1"
  >
    <p
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
          "textDecoration": "",
        }
      }
    >
      1
      .  
      Aprender React
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
  <li
    className="list-group-item"
    key="2"
  >
    <p
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
          "textDecoration": "",
        }
      }
    >
      2
      .  
      Aprender Mongo
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
</ul>
`;
